### PR TITLE
Subscriptions: RabbitMQ provider

### PR DESCRIPTION
**Pull request by Deathrage**
First version of RabbitMQ PubSub. Key assumptions are:
- Rabbit topics are not used
- HC topic is represented by a Rabbit exchange
- For every HC topic, each HC instance holds a Rabbit consumer (n topics and m instances -> n*m consumers)
- Every consumer has an exclusive queue to an exchange (n topics and m instances -> n\*m consumers -> n\*m queues)
- Single consumer consumes only one queue (queue is exclusive)
- Niether exchanges nor queues are durable
- Behaviour above can be changed directly in `Config` DTO and/or by overriding `IExchangeNameFactory` and `IQueueNameFactory`

Schema how messages are distributed (illustrates that `ISourceStream` does not directly connect to RabbitMQ):
image

### CODE REVIEW

**Comment by Deathrage** [comment]
@michaelstaib I have inspected the issue reported by code analyzer, it seemst that it is caused by tests in namespace `StrawberryShake.CodeGeneration.CSharp.Analyzers.CustomScalar` as there is 
```
public static class DependencyInjection
{
    public static void Configure(IServiceCollection services)
    {
        services.AddCustomScalarClient();
    }
}
```
But method `AddCustomScalarClient` does not exist. Is si supposed to get generated? If so, it does not happen.

**Comment by michaelstaib** [comment]
@Deathrage I am working on the code gen issue with strawberry shake ... so no worries. It started to happen when we moved to .net 7.

File: src/HotChocolate/Core/src/Subscriptions.RabbitMQ/ActiveConsumer.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
+using System;
+using RabbitMQ.Client.Events;
+
+namespace HotChocolate.Subscriptions.RabbitMQ;
+
+public class ActiveConsumer
+{
+    public AsyncEventingBasicConsumer Consumer { get; }
+
+    public ActiveConsumer(AsyncEventingBasicConsumer consumer, Action onEmpited)
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Subscriptions.RabbitMQ/ActiveConsumer.cs:10]: do you mean emitted? onEmpited?

Comment (Deathrage) [reply src/HotChocolate/Core/src/Subscriptions.RabbitMQ/ActiveConsumer.cs]: It's `onEmpited`, a callback called anytime consumer's source stream (resp. listen delegates) returns to 0.

Comment (michaelstaib) [reply src/HotChocolate/Core/src/Subscriptions.RabbitMQ/ActiveConsumer.cs]: The its OnEmptied

Comment (Deathrage) [reply src/HotChocolate/Core/src/Subscriptions.RabbitMQ/ActiveConsumer.cs]: it is I fixed it in commit 4e891d5

**Comment by michaelstaib** [comment]
We have reworked the subscription system and implemented this driver now with a separate PR.

**Pull request closed without merging**